hash/hash_comparator.go

```diff
@@ -28,45 +28,106 @@ import (
 
 var ErrUnknownHashAlgorithm = errors.New("unknown hash algorithm")
 
-func Compare(ctx context.Context, password []byte, hash []byte) error {
-	switch {
-	case IsBcryptHash(hash):
-		return CompareBcrypt(ctx, password, hash)
-	case IsArgon2idHash(hash):
-		return CompareArgon2id(ctx, password, hash)
-	case IsArgon2iHash(hash):
-		return CompareArgon2i(ctx, password, hash)
-	case IsPbkdf2Hash(hash):
-		return ComparePbkdf2(ctx, password, hash)
-	case IsScryptHash(hash):
-		return CompareScrypt(ctx, password, hash)
-	case IsSSHAHash(hash):
-		return CompareSSHA(ctx, password, hash)
-	case IsSHAHash(hash):
-		return CompareSHA(ctx, password, hash)
-	case IsFirebaseScryptHash(hash):
-		return CompareFirebaseScrypt(ctx, password, hash)
-	case IsMD5Hash(hash):
-		return CompareMD5(ctx, password, hash)
-	default:
-		return errors.WithStack(ErrUnknownHashAlgorithm)
+var hashComparators []PasswordHashComparator
```

Comment:
I don't think we need to refactor the package like this because there is already an established way to pass config to the hash comparators. Regardless of whether the code is nicer or not, in the end it makes it impossible to review with refactoring and new features mixed. Please use the existing way to pass the hash config, and revert all the refactoring. I am open to discuss that, but it should really be a separate PR.

Reply:
> there is already an established way to pass config to the hash comparators.

Maybe I missed something, but other than the PHC string, I don't see any way to pass conf data in the existing approach. IMHO the PHC string is not the right place for pepper values (see the discussion [here](https://github.com/ory/kratos/issues/2946#issuecomment-1453139768)). So refactoring is a precondition and not just to make the code nicer.